### ISSUE TITLE
Prototype Alpha

### ISSUE THREAD

**Issue by zfifteen** [zfifteen/dna-breathing-dynamics-encoding#29]
Prototype = one self-contained Python script you can drop into dna_breathing_dynamics_encoding/experiments/ or a gist and wire to your repo later.

1. Data + assumptions
	•	Input:
	•	sequences.fasta (or similar): headers = ids, sequences DNA (ACGTN…).
	•	labels.csv: columns id,label where label ∈ {0,1} (e.g. 1 = binding / edit, 0 = control).
	•	Biophysics (can refine later, but wired as explicit, editable tables):
	•	Breathing lifetimes [ms]: AT = 5, TA = 5, GC = 25, CG = 25; others via simple mapping / mean.
	•	Thermodynamic stability: per-bp ΔG in kcal/mol from a small dict (initially made-up placeholder, to be replaced by table/EPBD later).
	•	CZT:
	•	Window: full sequence or central window length L.
	•	Target harmonic: 10.5 bp/turn → spatial frequency f₀ = 1/10.5 cycles/bp.
	•	Band: narrow band around f₀ with M bins (e.g. ±10% relative bandwidth).

2. Signals and features

For each sequence:
	1.	Map each base pair to:
	•	τ_i = breathing lifetime (ms).
	•	ΔG_i = thermodynamic stability.
	2.	Build complex signal x_i = τ_i + 1j * ΔG_i.
	3.	Apply CZT about frequency f₀:
	•	Magnitude spectrum |X_k|.
	•	Phase arg(X_k).
	4.	Feature vector (per sequence), e.g.:
	•	mag_peak = max |X_k| in band.
	•	freq_argmax (index or effective frequency).
	•	phase_at_peak.
	•	band_energy = sum |X_k|² in band.
	•	Simple real/imag stats: mean(τ), var(τ), mean(ΔG), var(ΔG).
	5.	Baselines:
	•	Random encoding control: shuffle τ/ΔG across positions or use random real/imag from same marginal distributions and rerun CZT.
	•	GC-content feature: global %GC and maybe local %GC in sliding windows.

3. Minimal stats
	•	Split into train/test with fixed seed.
	•	Model: logistic regression (or linear SVM) on:
	•	(a) CZT breathing features only.
	•	(b) GC-only baseline.
	•	(c) Random encoding + CZT control.
	•	Metrics:
	•	ROC AUC on test.
	•	Cohen’s d for main scalar feature(s) (e.g. mag_peak) between label 0 vs 1.
	•	p-value from Mann–Whitney U or t-test for those features.
	•	Log everything deterministic: seed, file names, hyperparams, exact feature set.

4. Single-file prototype sketch

import numpy as np
import pandas as pd
from scipy.signal import czt
from scipy.stats import ttest_ind
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from Bio import SeqIO

RNG_SEED = 137
np.random.seed(RNG_SEED)

BREATH_LIFETIMES_MS = {
    "A": 5.0, "T": 5.0, "G": 25.0, "C": 25.0
}

DELTA_G = {
    "A": -1.0, "T": -1.0, "G": -2.0, "C": -2.0
}

def encode_sequence(seq: str):
    taus = []
    dgs = []
    for b in seq.upper():
        if b not in BREATH_LIFETIMES_MS:
            continue
        taus.append(BREATH_LIFETIMES_MS[b])
        dgs.append(DELTA_G[b])
    if not taus:
        return None
    x = np.array(taus) + 1j * np.array(dgs)
    return x

def gc_content(seq: str):
    s = seq.upper()
    gc = sum(b in "GC" for b in s)
    atgc = sum(b in "ATGC" for b in s)
    return gc / atgc if atgc > 0 else 0.0

def czt_band_features(x: np.ndarray, f0=1/10.5, rel_bw=0.1, m=64):
    n = len(x)
    # spatial frequency in cycles / bp mapped to digital radian freq
    w0 = 2 * np.pi * f0
    # center step and band
    w_start = w0 * (1 - rel_bw)
    w_end = w0 * (1 + rel_bw)
    k = np.arange(m)
    # chirp z: map line from w_start to w_end
    A = np.exp(1j * w_start)
    W = np.exp(1j * (w_end - w_start) / (m - 1))
    X = czt(x, m=m, w=W, a=A)
    mags = np.abs(X)
    idx = np.argmax(mags)
    mag_peak = mags[idx]
    band_energy = np.sum(mags**2)
    phase_peak = np.angle(X[idx])
    freq_eff = f0 * (1 - rel_bw + 2 * rel_bw * idx / (m - 1))
    return np.array([mag_peak, band_energy, phase_peak, freq_eff])

def random_encoding_control(x: np.ndarray):
    # shuffle real/imag separately
    r = x.real.copy()
    i = x.imag.copy()
    np.random.shuffle(r)
    np.random.shuffle(i)
    return r + 1j * i

def fasta_to_df(fasta_path: str, labels_csv: str):
    labels = pd.read_csv(labels_csv)
    label_map = dict(zip(labels["id"], labels["label"]))
    rows = []
    for rec in SeqIO.parse(fasta_path, "fasta"):
        seq_id = rec.id
        if seq_id not in label_map:
            continue
        rows.append({"id": seq_id, "seq": str(rec.seq), "label": label_map[seq_id]})
    return pd.DataFrame(rows)

def build_feature_table(df: pd.DataFrame):
    records = []
    for _, row in df.iterrows():
        seq = row["seq"]
        x = encode_sequence(seq)
        if x is None:
            continue
        czt_feat = czt_band_features(x)
        gc = gc_content(seq)
        tau = x.real
        dg = x.imag
        stats = np.array([
            tau.mean(), tau.var(), dg.mean(), dg.var(), gc
        ])
        # random-encoding CZT control
        xr = random_encoding_control(x)
        czt_rand = czt_band_features(xr)
        feat = np.concatenate([czt_feat, stats, czt_rand])
        records.append({
            "id": row["id"],
            "label": row["label"],
            "feat": feat,
            "mag_peak": czt_feat[0],
            "mag_peak_rand": czt_rand[0],
            "gc": gc
        })
    return records

def stack_features(records):
    X = np.stack([r["feat"] for r in records], axis=0)
    y = np.array([r["label"] for r in records])
    return X, y

def cohen_d(a, b):
    a = np.asarray(a)
    b = np.asarray(b)
    na, nb = len(a), len(b)
    va, vb = a.var(ddof=1), b.var(ddof=1)
    sp = np.sqrt(((na - 1)*va + (nb - 1)*vb) / (na + nb - 2))
    return (a.mean() - b.mean()) / sp if sp > 0 else 0.0

def run_experiment(fasta_path: str, labels_csv: str, test_size=0.3):
    df = fasta_to_df(fasta_path, labels_csv)
    records = build_feature_table(df)
    X, y = stack_features(records)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=test_size, random_state=RNG_SEED, stratify=y
    )
    clf = LogisticRegression(max_iter=1000)
    clf.fit(X_train, y_train)
    y_prob = clf.predict_proba(X_test)[:, 1]
    auc_czt = roc_auc_score(y_test, y_prob)

    # GC-only baseline
    gc_vals = np.array([r["gc"] for r in records]).reshape(-1, 1)
    Xg_tr, Xg_te, yg_tr, yg_te = train_test_split(
        gc_vals, y, test_size=test_size, random_state=RNG_SEED, stratify=y
    )
    clf_gc = LogisticRegression(max_iter=1000)
    clf_gc.fit(Xg_tr, yg_tr)
    y_prob_gc = clf_gc.predict_proba(Xg_te)[:, 1]
    auc_gc = roc_auc_score(yg_te, y_prob_gc)

    # scalar feature stats
    mag_peak = np.array([r["mag_peak"] for r in records])
    mag_peak_rand = np.array([r["mag_peak_rand"] for r in records])
    lbl = np.array([r["label"] for r in records])
    m1 = mag_peak[lbl == 1]
    m0 = mag_peak[lbl == 0]
    d = cohen_d(m1, m0)
    tstat, pval = ttest_ind(m1, m0, equal_var=False)
    # control: random encoding
    mr1 = mag_peak_rand[lbl == 1]
    mr0 = mag_peak_rand[lbl == 0]
    d_rand = cohen_d(mr1, mr0)
    tr, pval_rand = ttest_ind(mr1, mr0, equal_var=False)

    print("N =", len(records))
    print("AUC_CZT =", auc_czt)
    print("AUC_GC  =", auc_gc)
    print("Cohen_d(mag_peak, true) =", d, "p =", pval)
    print("Cohen_d(mag_peak_rand)  =", d_rand, "p =", pval_rand)

if __name__ == "__main__":
    import argparse
    ap = argparse.ArgumentParser()
    ap.add_argument("--fasta", required=True)
    ap.add_argument("--labels", required=True)
    args = ap.parse_args()
    run_experiment(args.fasta, args.labels)

This gives you: real-data only, complex breathing+ΔG encoding, CZT zoom around 10.5 bp/turn, baseline GC vs CZT breathing, and a random-encoding control, all with deterministic seed and minimal surface area; you can then swap the hard-coded biophysics for EPBD/pyDNA outputs without changing the pipeline shape.
